haskell/path_utils.bzl

```diff
@@ -83,3 +83,26 @@ def mk_name(ctx, name_prefix):
 
   """
   return "{0}-{1}-{2}".format(name_prefix, ctx.attr.name, ctx.attr.version)
+
+def mk_module_name(ctx, source_name, name_prefix):
+  """
+  Make a target-unique and `source_name`-unique name.
+
+  This is quite similar to `mk_name` but also uses a path built from
+  `source_name` to prevent clashes with other names produced using the same
+  `name_prefix`.
+
+  Args:
+    ctx: Rule context.
+    source_name: Source file name.
+    name_prefix: Template for the name.
+
+  Returns:
+    string: Target- and source-unique name.
+  """
+  return "{0}-{1}-{2}-{3}__".format(
```

Comment:
This is necessary so avoid prefixes when e.g. we have `Network.Socket` and `Network.Socket.Internal`. With underscores we have: `Network.Socket__` and `Network.Socket.Internal__` suffixes. Ugly hack is ugly.